### PR TITLE
MODNOTES-186 Rewrite tests for NoteLinks

### PR DESCRIPTION
**Pull request by oleksandr-dekin**
## Purpose
Added test case for links, update logic for removing note if note doesn't have any link

## Approach
Write test case for Links, added logic for removing notes if note doesn't contain link

## Learning
https://issues.folio.org/browse/MODNOTES-186

### CODE REVIEW

File: src/main/java/org/folio/notes/service/impl/NotesServiceImpl.java [hunk 1]
```diff
@@ -149,13 +149,20 @@ public void updateLinks(String objectType, String objectId, NoteLinkUpdateCollec
...
-    noteRepository.saveAll(noteEntities);
   }
 
   @Transactional
   @Override
   public void updateNote(UUID id, Note dto) {
+    if (dto.getLinks().isEmpty()) {
+      deleteNote(id);
+    }
     noteRepository.findById(id)
```
Comment (dmtkachenko) [inline src/main/java/org/folio/notes/service/impl/NotesServiceImpl.java:166]: Shouldn't this step be skipped if the note has been deleted by the previous `if {}` statement?

Comment (oleksandr-dekin) [reply src/main/java/org/folio/notes/service/impl/NotesServiceImpl.java]: You are right, changed

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 2]
```diff
@@ -243,11 +272,8 @@ void return422OnGetByIdWhenIdIsInvalid() throws Exception {
   @DisplayName("Update existing note")
   void updateExistingNoteType() throws Exception {
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:273]:
```suggestion
  void updateExistingNote() throws Exception {
```

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 3]
```diff
@@ -295,11 +322,695 @@ void return404OnDeleteByIdWhenItNotExist() throws Exception {
...
+    var secondResultNote = getNoteById(notes, secondNote.getId());
+    var thirdResultNote = getNoteById(notes, thirdNote.getId());
+
+    assertEquals(DEFAULT_LINK_AMOUNT + 1, firstResultNote.getLinks().size());
+    assertEquals(DEFAULT_LINK_AMOUNT, secondResultNote.getLinks().size());
+    assertEquals(DEFAULT_LINK_AMOUNT, thirdResultNote.getLinks().size());
+  }
+
+  @Test
+  @DisplayName("Remove an add links to notes")
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:369]:
```suggestion
  @DisplayName("Remove and add links to notes")
```

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 4]
```diff
@@ -295,11 +322,695 @@ void return404OnDeleteByIdWhenItNotExist() throws Exception {
...
+    var firsNoteWithAssignedLink = generateNote();
+    var secondNoteWithUnassignedLink = generateNote();
+
+    var content = getNoteLinks("/note-links/domain/" + NON_EXISTING_DOMAIN + "/type/" + PACKAGE_TYPE + "/id/" + PACKAGE_ID_1
+      + "?status=UNASSIGNED");
+    var notes = OBJECT_MAPPER.readValue(content, NoteCollection.class).getNotes();
+
+    assertEquals(0, notes.size());
+    assertEquals(DEFAULT_LINK_AMOUNT, firsNoteWithAssignedLink.getLinks().size());
+    assertEquals(DEFAULT_LINK_AMOUNT, secondNoteWithUnassignedLink.getLinks().size());
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:865]: Do we need these asserts there? I don't understand what they are checking

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 5]
```diff
@@ -295,11 +322,695 @@ void return404OnDeleteByIdWhenItNotExist() throws Exception {
...
+    var content = getNoteLinks("/note-links/domain/" + DOMAIN + "/type/" + PACKAGE_TYPE + "/id/" + PACKAGE_ID_1
+      + "?status=UNASSIGNED");
+    var notes = OBJECT_MAPPER.readValue(content, NoteCollection.class).getNotes();
+
+    assertEquals(1, notes.size());
+  }
+
+  @Test
+  @DisplayName("Should return list of notes with non existing domain")
+  void shouldReturnListOfNotesWithNonExistingDomain() throws Exception {
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:855]: Maybe should return an empty list of notes?

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 6]
```diff
@@ -295,11 +322,695 @@ void return404OnDeleteByIdWhenItNotExist() throws Exception {
...
+  @Test
+  @DisplayName("Should return list of notes with status all")
+  void shouldReturnListOfNotesWithStatusAll() throws Exception {
+    generateNote();
+    generateNote();
+
+    var content = getNoteLinks("/note-links/domain/" + DOMAIN + "/type/" + PACKAGE_TYPE + "/id/" + PACKAGE_ID_1 + "?status=all");
+    var notes = OBJECT_MAPPER.readValue(content, NoteCollection.class).getNotes();
+
+    assertEquals(2, notes.size());
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:877]: Maybe will be better to add a link to one of the notes to check that all notes are returned: assigned and not assigned?

File: src/test/java/org/folio/notes/controller/NoteControllerTest.java [hunk 7]
```diff
@@ -295,11 +322,695 @@ void return404OnDeleteByIdWhenItNotExist() throws Exception {
...
+      "/note-links/domain/" + DOMAIN + "/type/" + PACKAGE_TYPE
+        + "/id/" + PACKAGE_ID_1 + "?order=wrong")
+      .headers(defaultHeaders()))
+      .andExpect(status().isUnprocessableEntity())
+      .andExpect(errorMessageMatch(containsString("Failed to convert value of type 'java.lang.String' to required type 'org.folio.notes.domain.dto.OrderDirection")));
+  }
+
+  @Test
+  @DisplayName("Should return note list when search by title and note type")
+  void shouldReturn400WithErrorMessageWrongLimitAndOffset() throws Exception {
```
Comment (psmagin) [inline src/test/java/org/folio/notes/controller/NoteControllerTest.java:916]: Display name is not consistent with method name.

**Review by psmagin — APPROVED** [review]

**Review by Shans-Kaluhin — APPROVED** [review]

**Pull request merged**